File: desktop-app/src/store/projectStore.ts
import { create } from 'zustand'
import { Project, ProjectCreate } from '../types/project'
import * as api from '../api/endpoints'

interface ProjectStore {
  projects: Project[]
  loading: boolean
  error: string | null

  fetchProjects: () => Promise<void>
  createProject: (data: ProjectCreate) => Promise<Project | null>
  updateProject: (id: number, data: ProjectCreate) => Promise<void>
  deleteProject: (id: number) => Promise<void>
  // *Actualiza una URL de imagen en el estado local tras subirla a Cloudinary
  setImage: (id: number, field: 'hero_image_url' | 'card_thumbnail_url', url: string) => void
}

const useProjectStore = create<ProjectStore>((set) => ({
  projects: [],
  loading: false,
  error: null,

  // *Carga todos los proyectos desde la API (sin filtros — vista admin)
  fetchProjects: async () => {
    set({ loading: true, error: null })
    try {
      const response = await api.getProjects()
      set({ projects: response.data, loading: false })
    } catch {
      set({ error: 'Error al cargar proyectos', loading: false })
    }
  },

  // *Crea un proyecto y lo agrega al array local
  createProject: async (data) => {
    set({ loading: true, error: null })
    try {
      const response = await api.createProject(data)
      set((state) => ({
        projects: [...state.projects, response.data],
        loading: false,
      }))
      return response.data
    } catch {
      set({ error: 'Error al crear proyecto', loading: false })
      return null
    }
  },

  // *Actualiza un proyecto en la API y en el array local
  updateProject: async (id, data) => {
    set({ loading: true, error: null })
    try {
      const response = await api.updateProject(id, data)
      set((state) => ({
        projects: state.projects.map((p) => (p.id === id ? response.data : p)),
        loading: false,
      }))
    } catch {
      set({ error: 'Error al actualizar proyecto', loading: false })
    }
  },

  // *Actualiza una URL de imagen en el estado local sin hacer otra llamada a la API
  setImage: (id, field, url) => {
    set((state) => ({
      projects: state.projects.map((p) =>
        p.id === id ? { ...p, [field]: url } : p,
      ),
    }))
  },

  // *Elimina un proyecto de la API y del array local
  deleteProject: async (id) => {
    set({ loading: true, error: null })
    try {
      await api.deleteProject(id)
      set((state) => ({
        projects: state.projects.filter((p) => p.id !== id),
        loading: false,
      }))
    } catch {
      set({ error: 'Error al eliminar proyecto', loading: false })
    }
  },
}))

export default useProjectStore

File: desktop-app-flutter/lib/state/project_store.dart
import 'package:flutter/foundation.dart';
import '../api/project_api.dart';
import '../models/project.dart';

/// Mirror del store de Zustand (projectStore.ts): misma forma,
/// mismos métodos, mismo comportamiento (incluida la actualización local
/// de imagen sin refetch tras un upload exitoso).
class ProjectStore extends ChangeNotifier {
  final ProjectApi _api;
  ProjectStore(this._api);

  List<Project> projects = [];
  bool loading = false;
  String? error;

  Future<void> fetchProjects() async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      projects = await _api.getProjects();
    } catch (_) {
      error = 'Error al cargar proyectos';
    }
    loading = false;
    notifyListeners();
  }

  Future<Project?> createProject(Map<String, dynamic> data) async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      final created = await _api.createProject(data);
      projects = [...projects, created];
      loading = false;
      notifyListeners();
      return created;
    } catch (_) {
      error = 'Error al crear proyecto';
      loading = false;
      notifyListeners();
      return null;
    }
  }

  Future<void> updateProject(int id, Map<String, dynamic> data) async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      final updated = await _api.updateProject(id, data);
      projects = [for (final p in projects) p.id == id ? updated : p];
    } catch (_) {
      error = 'Error al actualizar proyecto';
    }
    loading = false;
    notifyListeners();
  }

  Future<void> deleteProject(int id) async {
    loading = true;
    error = null;
    notifyListeners();
    try {
      await _api.deleteProject(id);
      projects = projects.where((p) => p.id != id).toList();
    } catch (_) {
      error = 'Error al eliminar proyecto';
    }
    loading = false;
    notifyListeners();
  }

  /// Patch local de la URL de imagen tras un upload exitoso — sin refetch,
  /// igual que setImage() en el store original.
  void setImage(int id, String field, String url) {
    projects = [
      for (final p in projects)
        if (p.id == id) _withImage(p, field, url) else p,
    ];
    notifyListeners();
  }

  Project _withImage(Project p, String field, String url) {
    final copy = p.deepCopy();
    if (field == 'hero_image_url') {
      copy.heroImageUrl = url;
    } else if (field == 'card_thumbnail_url') {
      copy.cardThumbnailUrl = url;
    }
    return copy;
  }
}
